src/Microsoft.DotNet.Interactive/KernelExtensions.cs

```diff
@@ -208,6 +209,194 @@ public static T UseLogMagicCommand<T>(this T kernel)
     }
 
+    public static T UseSet<T>(this T kernel) where T : Kernel
+    {
+        var fromResultOption = new Option<bool>(
+            "--from-result",
+            description:"Captures the execution result.");
+
+
+        var fromValueOption = new Option<string>(
+            "--from-value",
+            description: "Specifies a value to be stored directly. Specifying @input:value allows you to prompt the user for this value.",
+            parseArgument: result =>
+            {
+                if (SetErrorIfAlsoUsed(fromResultOption, result))
+                {
+                    return null;
+                }
+
+                return result.Tokens.Single().Value;
+            });
+
+        var mimeTypeOption = new Option<string>("--mime-type", "Share the value as a string formatted to the specified MIME type.")
+            .AddCompletions(
+                JsonFormatter.MimeType,
+                HtmlFormatter.MimeType,
+                PlainTextFormatter.MimeType);
+
+        var nameOption = new Option<string>(
+            "--name",
+            description: "This is the name used to declare and set the value in the kernel.",
+            parseArgument: result =>
+            {
+                if (SetErrorIfNoneAreUsed(result, fromResultOption, fromValueOption))
+                {
+                    return null;
+                }
+
+                return result.Tokens.Single().Value;
+            }
+        )
+        {
+            IsRequired = true
+        };
+
+        var set = new Command("#!set")
+        {
+            nameOption,
+            fromResultOption,
+            fromValueOption,
+        };
+
+        set.SetHandler(cmdLineContext =>
+        {
+            HandleSetMagicCommand(kernel, cmdLineContext, fromResultOption, nameOption, fromValueOption);
+        });
+
+        kernel.AddDirective(set);
+
+        return kernel;
+
+        bool SetErrorIfAlsoUsed(Option otherOption, ArgumentResult result)
+        {
+            var otherOptionResult = result.FindResultFor(otherOption);
+
+            if (otherOptionResult is { })
+            {
+                result.ErrorMessage =
+                    $"The {otherOptionResult.Token.Value} and {((OptionResult)result.Parent).Token.Value} options cannot be used together.";
+
+                return true;
+            }
+
+            return false;
+        }
+
+        bool SetErrorIfNoneAreUsed( ArgumentResult result, params Option[] options)
+        {
+            var otherOptionResults = options.Select( result.FindResultFor).Where(option => option is not null).ToList();
+
+            if (otherOptionResults.Count == 0)
+            {
+                result.ErrorMessage =
+                    $"At least one of the options [{string.Join(", ", options.Select(o => o.Name))}] must be specified.";
+
+                return true;
+            }
+
+            return false;
+        }
+    }
+
+    private static void HandleSetMagicCommand<T>(
```

Comment:
This method should be leveraging and/or refactoring the existing `GetValueAndSendTo` method.

`#!share --from csharp x` should be the same code path (after interpolation is done) as `#!set --name x --value @csharp:x`.